src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Collections.Generic;
+using Microsoft.AspNet.Mvc.OptionDescriptors;
+
+namespace Microsoft.AspNet.Mvc
+{
+    /// <summary>
+    /// Extension methods for collections of option descriptors.
+    /// </summary>
+    public static class OptionDescriptorExtensions
+    {
+        /// <summary>
+        /// Returns the only instance of <typeparamref name="TInstance"/> from a sequence of option descriptors.
```

Comment:
Because that's the design we asked him to implement. This is a code review and not a design review. 

Design wise:
The pattern follows normal .net API design.

Most users would use the overload that throw. Because it works as expected most of the time.

3rd party library developers will use the or default overload because they cannot make assumptions, or they will write their own method.

-----Original Message-----
From: "Doug Bunting" <notifications@github.com>
Sent: ‎2/‎7/‎2015 4:32 PM
To: "aspnet/Mvc" <Mvc@noreply.github.com>
Cc: "Yishai Galatzer" <yishai_galatzer@yahoo.com>
Subject: Re: [Mvc] [Fix for #1787] Add Extension Methods toIEnumerable<IOptionDescriptor<object>> (#1906)

In src/Microsoft.AspNet.Mvc.Core/OptionDescriptors/OptionDescriptorExtensions.cs:
> +﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
> +// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
> +
> +using System.Linq;
> +using System.Collections.Generic;
> +using Microsoft.AspNet.Mvc.OptionDescriptors;
> +
> +namespace Microsoft.AspNet.Mvc
> +{
> +    /// <summary>
> +    /// Extension methods for collections of option descriptors.
> +    /// </summary>
> +    public static class OptionDescriptorExtensions
> +    {
> +        /// <summary>
> +        /// Returns the only instance of <typeparamref name="TInstance"/> from a sequence of option descriptors.
InstanceOf() is supposed to throw, and InstanceOfOrDefault is that one that returns null
@yishaigalatzer why do we need both?
also both methods throw when multiple elements in the list are of the given type. is that desired?
generally, my expectation was for ...InstanceOf<TInstance>() to be shorthand for ...Select(descriptor => descriptor.Instance).OfType<TInstance>().FirstOrDefault(); ending with ...First() if throwing a system-provided InvalidOperationException was preferred.
—
Reply to this email directly or view it on GitHub.